### ISSUE TITLE
Refactoring  IStackedAdapter

### ISSUE THREAD

**Issue by IanMayo** [debrief/limpet#381]
## Issue

To integrate our StackedCharts View with separate applications (Debrief & Limpet) we let the applications register adapters.  Then, when a data object is dropped into the Stacked Chart View (in edit mode), the Stacked Charts processing then loops through the adapters to see if any can convert it.

Here is the interface that our registered adapters meet:

```
public interface IStackedAdapter
{
  boolean canConvert(Object data);  
  List<Object> convert(Object data);
}
```

Our stacked charts UI asks the registered  adapters to examine the data object, and return a list of all of the Model items it can produce (so far this is Datasets or ScatterSets).

But,  this introduces a problem:  it's possible that a dataset could be converted into a Dataset or a ScatterSet.

The solution I've used so far is for the adapter to ask the user whether the user wants a Dataset or a ScatterSet.

But, we now have a ScatterSet landing pad.  So our stacked charts UI knows that it wants the dropped object to be converted into a ScatterSet.

I can think of the following solutions.
## 1. Specify the required data type

```
public interface IStackedAdapter1
{
  boolean canConvert(Object data, class targetType);  
  List<Object> convert(Object data, class targetType);
}
```
### Benefits
- future-proof: we can introduce new targetType objects without an API change
### Problems
- the API isn't specific, the targetType could be anything, so we need a large if/else to process it
- the return type isn't specific, so our UI processing need to check the received data object.
## 2. Multiple converters

```
public interface IStackedAdapter2
{
  boolean canConvertToDataset(Object data);  
  List<Dataset> convertToDataset(Object data);

  boolean canConvertToScatterSet(Object data);  
  List<ScatterSet> convertToScatterSet(Object data);
}
```
### Benefits
- explicit API, we can rely on the type of model object that is returned
### Problems
- untidy, the API is covering multiple separate conversions
- if we add a future type, existing adapters will have to be modified to meet the new Interface
## 3. Multiple Adapters

```
public interface IStackedDatasetAdapter
{
  boolean canConvertToDataset(Object data);  
  List<Dataset> convertToDataset(Object data);
}
public interface IStackedScatterssetAdapter
{
  boolean canConvertToScatterSet(Object data);  
  List<ScatterSet> convertToScatterSet(Object data);
}

```
### Benefits
- explicit API, we can rely on the type of model object that is returned
- tidy expansion plan for future types, we simply add another adapter.
### Problems
- our StackedCharts processing will be slightly more complex (since we're handing/registering multiple types of adapter)

**Comment by IanMayo** [comment]
@dinkoivanov @theanuradha - guys, I'd appreciate your thoughts on the above please.

Am I missing anything?  Do you have any preferences/tips on which is the best option?

**Comment by theanuradha** [comment]
@IanMayo I go with Option one [1]  with Changes

```
public interface IStackedAdapter1
{
  <T> boolean canConvert(Object data, TargetType<T> targetType);  
  <T> List<T> convert(Object data, TargetType<T> targetType);
}
```

```
public class TargetType<T>
{
  private TargetType()
  {
   //hide
  }

  public static final TargetType<ScatterSet> SCATTERSET = new TargetType<ScatterSet>();
  public static final TargetType<DataSet> DATASET = new TargetType<DataSet>();
}
```

**Comment by IanMayo** [comment]
@theanuradha - that's interesting.  Could you explain the changes?  I mostly understand the generics, but can you explain the purpose of the private constructor?

Can you give me a larger code sample?

**Comment by theanuradha** [comment]
@IanMayo TargetType I added as **Enums** DO not allow generics so used private constructor to prevent TargetType create outside. So it act as a Enum with generics. So code is type safe to outside of api use .
this allow api user to do. 
`List< ScatterSet > scatterSet =  adapter.convert(obj, TargetType. SCATTERSET);`

this remove the issues on option 1 

> Problems
> 
> the API isn't specific, the targetType could be anything, so we need a large if/else to process it
> the return type isn't specific, so our UI processing need to check the received data object.

**Comment by theanuradha** [comment]
this is impl of adapter look like 

```
public class ScatterSetAdapterImpl implements IStackedAdapter1
{

  @Override
  public <T> boolean canConvert(Object data, TargetType<T> targetType)
  {
    if(TargetType.SCATTERSET==targetType)
    {
      //we support SCATTERSET
      //other checks on passed object
      return true;
    }
    return false;
  }

  @SuppressWarnings("unchecked")
  @Override
  public <T> List<T> convert(Object data, TargetType<T> targetType)
  {
    if(TargetType.SCATTERSET==targetType)
    {
      List<ScatterSet> scatterSets = new ArrayList<ScatterSet>();
      //we support SCATTERSET
      //other checks convet process
      return (List<T>)scatterSets;

    }
    throw new IllegalArgumentException("not suppoted target type: "+targetType);
  }

}
```

**Comment by IanMayo** [comment]
The interface is used in an RCP extension: https://raw.githubusercontent.com/debrief/limpet/stacked_charts_poc/info.limpet.stackedcharts.ui/schema/stacked_adapter.exsd

So, I'm nervous about the generics in this RCP extension.

```
the API isn't specific, the targetType could be anything, so we 
     need a large if/else to process it
```

Your code still requires the if/else construct.

**Comment by theanuradha** [comment]
@IanMayo  targetType we limit API user with only possible types. so they can not send anything. Yes  if/else will be limited as it depend on  number of types  TargetType provide.

RCP extension will not affect as only methods has generics than defined on interface level.

**Comment by IanMayo** [comment]
Fixed in #383
